File: src/content/docs/topics/cs/coding-problems/graphs/207-course-schedule-approach2.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function canFinish(numCourses: number, prerequisites: number[][]): boolean {
    const graph: number[][] = Array.from({ length: numCourses }, () => []);
    for (const [a, b] of prerequisites) graph[b].push(a);

    const WHITE = 0, GRAY = 1, BLACK = 2;
    const color = new Array(numCourses).fill(WHITE);

    function dfs(n: number): boolean {
        if (color[n] === GRAY) return false;
        if (color[n] === BLACK) return true;
        color[n] = GRAY;
        for (const nb of graph[n]) {
            if (!dfs(nb)) return false;
        }
        color[n] = BLACK;
        return true;
    }

    for (let c = 0; c < numCourses; c++) {
        if (!dfs(c)) return false;
    }
    return true;
}

assert(canFinish(2, [[1, 0]]) === true);
assert(canFinish(2, [[1, 0], [0, 1]]) === false);
assert(canFinish(5, []) === true);
assert(canFinish(1, []) === true);
assert(canFinish(3, [[1, 0], [2, 1], [0, 2]]) === false);
assert(canFinish(4, [[1, 0], [2, 0], [3, 1], [3, 2]]) === true);
console.log('all tests pass');

File: src/content/docs/topics/cs/coding-problems/graphs/207-course-schedule-approach2.swift
// LEETCODE_TYPE: Solution
func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}
final class Solution {
    func canFinish(_ numCourses: Int, _ prerequisites: [[Int]]) -> Bool {
        var graph = Array(repeating: [Int](), count: numCourses)
        for edge in prerequisites { graph[edge[0]].append(edge[1]) }
        var color = Array(repeating: 0, count: numCourses)
        func visit(_ course: Int) -> Bool {
            if color[course] == 1 { return false }
            if color[course] == 2 { return true }
            color[course] = 1
            for next in graph[course] where !visit(next) { return false }
            color[course] = 2
            return true
        }
        return (0..<numCourses).allSatisfy(visit)
    }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:ed8fd2c6554df64a34728495c5a678faa1d9c8dfd7659ef5258e965ff11d7621
    expectEqual(Solution().canFinish(3, [[1, 0], [2, 1]]), true, "simple-chain")
    expectEqual(Solution().canFinish(2, [[1, 0], [0, 1]]), false, "two-course-cycle")
    expectEqual(Solution().canFinish(1, []), true, "one-course")
    // EXCLUDED_VECTOR course-out-of-range: [2,[[2,0]]] | Every course identifier must be smaller than numCourses.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
